File: frontend/src/nutrition/body.ts
// 하루 칼로리 목표 계산(4b-2 Task 6). 브라우저 API 없음, 오늘은 인자로 받는다.
import type { Activity, BodyGoal, BodyProfile, Sex } from "../api";

export const SEXES: [Sex, string][] = [
  ["female", "여성"],
  ["male", "남성"],
];

export const ACTIVITIES: { value: Activity; label: string; hint: string; factor: number }[] = [
  { value: "sedentary", label: "거의 없음", hint: "대부분 앉아서 지내요", factor: 1.2 },
  { value: "light", label: "가벼움", hint: "주 1~3번 가볍게 걷거나 운동해요", factor: 1.375 },
  { value: "moderate", label: "보통", hint: "주 3~5번 운동해요", factor: 1.55 },
  { value: "active", label: "많음", hint: "주 6~7번 운동해요", factor: 1.725 },
  { value: "very_active", label: "매우 많음", hint: "매일 힘든 운동이나 몸을 많이 쓰는 일을 해요", factor: 1.9 },
];

export const GOALS: [BodyGoal, string][] = [
  ["maintain", "유지"],
  ["lose", "감량"],
  ["gain", "증량"],
];

export const MIN_AGE = 20;
export const MAX_AGE = 99;
export const MAX_TARGET = 5000;

export const goalLabel = (goal: BodyGoal) => GOALS.find(([g]) => g === goal)![1];
export const kcalNumber = (n: number) => Math.round(n).toLocaleString("ko-KR");

/** 연 나이(결정 1): "2026-09-15", 1994 → 32 */
export const ageOf = (birthYear: number, today: string) => Number(today.slice(0, 4)) - birthYear;

export interface DailyTarget {
  bmr: number;
  tdee: number;
  target: number;
  floored: boolean;
  capped: boolean;
  factor: number;
}

/** Mifflin–St Jeor × 활동계수, 목표(결정 3). 모두 정수 반올림 */
export function dailyTarget(
  p: Pick<BodyProfile, "sex" | "birth_year" | "height_cm" | "weight_kg" | "activity" | "goal">,
  today: string,
): DailyTarget {
  const raw = 10 * p.weight_kg + 6.25 * p.height_cm - 5 * ageOf(p.birth_year, today) + (p.sex === "male" ? 5 : -161);
  const factor = ACTIVITIES.find((a) => a.value === p.activity)!.factor;
  const bmr = Math.round(raw);
  const tdee = Math.round(raw * factor);
  if (p.goal === "lose") return { bmr, tdee, target: Math.max(tdee - 500, bmr), floored: tdee - 500 < bmr, capped: false, factor };
  if (p.goal === "gain")
    return { bmr, tdee, target: Math.min(tdee + 500, MAX_TARGET), floored: false, capped: tdee + 500 > MAX_TARGET, factor };
  return { bmr, tdee, target: tdee, floored: false, capped: false, factor };
}

/** 결과 상자 아래 설명(시안 BODY SHEET). 모든 경우 끝에 참고용 문구 */
export function targetNote(t: DailyTarget, goal: BodyGoal): string {
  const tail = `Mifflin–St Jeor 식 × 활동계수 ${t.factor} · 의료 조언이 아니라 참고용이에요.`;
  if (goal === "lose") return `${t.floored ? "필요량 − 500kcal이지만 기초대사량보다 낮게는 제안하지 않아요." : "필요량 − 500kcal이에요."} ${tail}`;
  if (goal === "gain") return `${t.capped ? "필요량 + 500kcal이지만 5,000kcal까지만 제안해요." : "필요량 + 500kcal이에요."} ${tail}`;
  return tail;
}

/** 입력 글자 → 저장할 값. 틀리면 칸별 안내 문구(화면 아래 줄), 비었으면 null 값 */
export function parseProfileInput(
  f: { sex: Sex | null; birthYear: string; height: string; weight: string; activity: Activity | null; goal: BodyGoal },
  today: string,
): { value: Omit<BodyProfile, "updated_at"> | null; errors: { birthYear?: string; height?: string; weight?: string } } {
  const year = Number(today.slice(0, 4));
  const num = (s: string) => (s.trim() === "" ? null : Number(s.replace(",", ".")));
  const birth = num(f.birthYear),
    height = num(f.height),
    weight = num(f.weight);
  const errors: { birthYear?: string; height?: string; weight?: string } = {};
  if (birth !== null && !(Number.isInteger(birth) && birth >= year - MAX_AGE && birth <= year - MIN_AGE))
    errors.birthYear = `${year - MAX_AGE}~${year - MIN_AGE}년 사이로 입력해주세요`;
  if (height !== null && !(height >= 120 && height <= 230)) errors.height = "120~230cm 사이로 입력해주세요";
  if (weight !== null && !(weight >= 30 && weight <= 250)) errors.weight = "30~250kg 사이로 입력해주세요";
  const complete = f.sex && f.activity && birth !== null && height !== null && weight !== null && !Object.keys(errors).length;
  return {
    value: complete
      ? { sex: f.sex!, birth_year: birth!, height_cm: Math.round(height! * 10) / 10, weight_kg: Math.round(weight! * 10) / 10, activity: f.activity!, goal: f.goal }
      : null,
    errors,
  };
}

File: backend/app/body.py
import math

from flask import Blueprint, abort, g, jsonify, request

from .auth import login_required
from .ingredients import seoul_today
from .models import BodyProfile, db
from .validation import commit_or_duplicate, integer, iso_datetime

# 몸 정보(스펙 21절, 4b-2). 하루 목표 kcal 자체는 저장하지 않고 화면이 계산한다(결정 2, src/nutrition/body.ts).
bp = Blueprint("body", __name__, url_prefix="/api")

SEXES = ("female", "male")
ACTIVITIES = ("sedentary", "light", "moderate", "active", "very_active")
GOALS = ("maintain", "lose", "gain")
MIN_AGE, MAX_AGE = 20, 99  # 연 나이. Mifflin–St Jeor는 성인 공식(결정 1)
SAVE_RACE = "방금 저장했어요. 다시 불러와주세요."


def _number(value, label, lo, hi):
    """bool이 아닌 lo~hi 숫자, 소수 첫째 자리로 반올림."""
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value) or not lo <= value <= hi:
        abort(400, f"{label} {lo}~{hi} 사이 숫자로 입력해주세요.")
    return round(float(value), 1)


def profile_json(profile):
    return {
        "sex": profile.sex, "birth_year": profile.birth_year, "height_cm": profile.height_cm,
        "weight_kg": profile.weight_kg, "activity": profile.activity, "goal": profile.goal,
        "updated_at": iso_datetime(profile.updated_at),
    }


def _my_profile():
    return BodyProfile.query.filter_by(user_id=g.user.id).first()


@bp.get("/body-profile")
@login_required
def get_body_profile():
    profile = _my_profile()
    res = jsonify(profile=profile_json(profile) if profile else None)
    res.headers["Cache-Control"] = "no-store"  # 건강 정보
    return res


@bp.put("/body-profile")
@login_required
def put_body_profile():
    data = request.get_json(silent=True)
    if not isinstance(data, dict):
        abort(400, "잘못된 요청이에요.")
    sex = data.get("sex")
    if sex not in SEXES:
        abort(400, "성별을 골라주세요.")
    this_year = seoul_today().year
    birth_year = integer(data.get("birth_year"), "태어난 해는", this_year - MAX_AGE, this_year - MIN_AGE)
    height_cm = _number(data.get("height_cm"), "키는", 120, 230)
    weight_kg = _number(data.get("weight_kg"), "몸무게는", 30, 250)
    activity = data.get("activity")
    if activity not in ACTIVITIES:
        abort(400, "활동량을 골라주세요.")
    goal = data.get("goal")
    if goal not in GOALS:
        abort(400, "목표를 골라주세요.")

    profile = _my_profile()
    if profile is None:
        profile = BodyProfile(user_id=g.user.id)
        db.session.add(profile)
    profile.sex, profile.birth_year = sex, birth_year
    profile.height_cm, profile.weight_kg = height_cm, weight_kg
    profile.activity, profile.goal = activity, goal
    commit_or_duplicate(SAVE_RACE)  # 처음 저장이 동시에 두 번 오면 UNIQUE(user_id) 충돌
    return jsonify(profile=profile_json(profile))


@bp.delete("/body-profile")
@login_required
def delete_body_profile():
    profile = _my_profile()
    if profile is not None:
        db.session.delete(profile)
        db.session.commit()
    return "", 204
